google-cloud-spanner/src/test/java/com/google/cloud/spanner/it/ITCommitTimestampTest.java

```diff
@@ -0,0 +1,319 @@
+import com.google.cloud.spanner.IntegrationTestEnv;
+import com.google.cloud.spanner.Key;
+import com.google.cloud.spanner.Mutation;
+import com.google.cloud.spanner.SpannerExceptionFactory;
+import com.google.cloud.spanner.Struct;
+import com.google.cloud.spanner.TimestampBound;
+import com.google.cloud.spanner.Value;
+import com.google.cloud.spanner.testing.RemoteSpannerHelper;
+import com.google.common.collect.ImmutableList;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.concurrent.ExecutionException;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Integration test for commit timestamp of Cloud Spanner. */
+@Category(IntegrationTest.class)
+@RunWith(JUnit4.class)
+public class ITCommitTimestampTest {
+  @ClassRule public static IntegrationTestEnv env = new IntegrationTestEnv();
+  @Rule public final ExpectedException expectedException = ExpectedException.none();
+  private Database db;
+  private DatabaseClient client;
+  private DatabaseAdminClient dbAdminClient;
+  private RemoteSpannerHelper testHelper;
+  private String instanceId;
+  private String databaseId;
+
+  @Before
+  public void setUp() throws Exception {
+    testHelper = env.getTestHelper();
+    db =
+        testHelper.createTestDatabase(
+            "CREATE TABLE T ("
+                + "K  STRING(MAX) NOT NULL,"
+                + "T1   TIMESTAMP OPTIONS (allow_commit_timestamp = true),"
+                + "T2   TIMESTAMP OPTIONS (allow_commit_timestamp = true),"
+                + "T3   TIMESTAMP,"
+                + ") PRIMARY KEY (K)");
+    client = testHelper.getDatabaseClient(db);
+    dbAdminClient = testHelper.getClient().getDatabaseAdminClient();
+    instanceId = testHelper.getInstanceId().getInstance();
+    databaseId = db.getId().getDatabase();
+  }
+
+  private Timestamp write(Mutation m) {
+    return client.write(Arrays.asList(m));
+  }
+
+  private Struct readRow(DatabaseClient client, String table, Key key, String... columns) {
+    return client.singleUse(TimestampBound.strong()).readRow(table, key, Arrays.asList(columns));
+  }
+
+  @Test
+  public void writeCommitTimestamp() {
+    // 1. timestamps auto populated and returned should be the same
+    Timestamp commitTimestamp =
+        write(
+            Mutation.newInsertOrUpdateBuilder("T")
+                .set("K")
+                .to("a")
+                .set("T1")
+                .to(Value.COMMIT_TIMESTAMP)
+                .set("T2")
+                .to(Value.COMMIT_TIMESTAMP)
+                .build());
+    Struct row = readRow(client, "T", Key.of("a"), "T1", "T2");
+    assertThat(row.getTimestamp(0)).isEqualTo(commitTimestamp);
+    assertThat(row.getTimestamp(1)).isEqualTo(commitTimestamp);
```

Comment:
Could we also add an assertion that the commit timestamp returned by the `write(...)` method is equal to the commit timestamp written to these columns?

Reply:
Apologies if I missed something here, but line number 102 and 103 are doing exactly that. They are checking that commitTimestamp returned by write method on line 91 is same as that written to the Key "a" for columns "T1" and "T2".

Reply:
Sorry, you're completely right, I have no idea where my mind was when I added that comment :-)